apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs

```diff
@@ -0,0 +1,109 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Api.Gax.Grpc.RetrySettings;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// Helper class for retrying Spanner operations.
+    /// Note that this is not the same as the retrying mechanism for each
+    /// of the RPCs, but for a higher level kind of operation that may include
+    /// one or several RPCs.
+    /// A good example of a higher level operation is the retrying of a 
+    /// transaction commit on abort, which requieres the retrying of all 
+    /// the transaction work and of the commit.
+    /// TODO(atarafamas, jskeet): Consider making public or entirely eliminating once we know 
+    /// what we can reuse from GAX retry helpers.
+    /// </summary>
+    internal static class SpannerRetryHelpers
+    {
+        internal static async Task<TResult> ExecuteWithRetryAsync<TResult>(Func<Task<TResult>> fn, SpannerRetryOptions retryOptions, IClock clock, IScheduler scheduler, CancellationToken cancellationToken)
+        {
+            if (retryOptions == null)
+            {
+                return await fn().ConfigureAwait(false);
+            }
+
+            DateTime start = clock.GetCurrentDateTimeUtc();
+            DateTime? overallDeadline = retryOptions.OverallTimeout.CalculateDeadline(clock);
+            IJitter jitter = new MinimumBoundJitter(retryOptions.DelayJitterMultiplier);
+            TimeSpan retryDelay = retryOptions.BackoffSettings.Delay;
+            while (true)
+            {
+                try
+                {
+                    var result = await fn().ConfigureAwait(false);
+                    return result;
+                }
+                catch (SpannerException e) when (retryOptions.ExceptionFilter(e))
+                {
+                    // If there's a recommended retry delay specified on the exception
+                    // we should at least wait for that.
+                    if (e.RecommendedRetryDelay.HasValue && e.RecommendedRetryDelay.Value > retryDelay)
+                    {
+                        retryDelay = e.RecommendedRetryDelay.Value;
+                    }
+                    TimeSpan actualDelay = jitter.GetDelay(retryDelay);
+                    DateTime expectedRetryTime = clock.GetCurrentDateTimeUtc() + actualDelay;
+                    if (expectedRetryTime > overallDeadline)
+                    {
+                        throw;
+                    }
+                    await scheduler.Delay(actualDelay, cancellationToken).ConfigureAwait(false);
+                    retryDelay = retryOptions.BackoffSettings.NextDelay(retryDelay);
+                }
+            }
+        }
+
+        private class MinimumBoundJitter : IJitter
```

Comment:
I don't think it's a good idea to implement `IJitter` here. It doesn't fulfill the same contract, as is evident by the fact that the parameter is `minDelay` whereas in `IJitter` it's `maxDelay`. I think just a simple method (passing in `retryOptions.DelayJitterMultiplier`) would be better. Apologies for having missed this before.